### PR TITLE
[SourceKit] Merge local rename and related identifiers

### PR DESCRIPTION
**Pull request by ahoppen**
The related identifiers request was sufficiently similar to local rename that we can implement it in terms of local rename.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

File: lib/Refactoring/LocalRename.cpp [hunk 1]
```diff
@@ -283,21 +299,30 @@ swift::refactoring::getRenameInfo(ResolvedCursorInfoPtr cursorInfo) {
 }
 
 class RenameRangeCollector : public IndexDataConsumer {
+  StringRef USR;
+  StringRef newName;
+  std::unique_ptr<StringScratchSpace> stringStorage;
+  std::vector<RenameLoc> locations;
```
Comment (bnbarham) [inline lib/Refactoring/LocalRename.cpp:305]: Is the collector actually used again with `newName`, or do we always end up doing the syntactic rename after we have locations?

**Comment by ahoppen** [comment]
While looking at open issues regarding related idents and local rename, we found that there are still a few issues here, one of them is that `NameMatcher` currently assumes that it’s iterating in source order while in-fact it isn’t because it’s an ASTVisitor. This causes us to skip references that have a postfix operator after them.

The proper solution here is to write `NameMatcher` in terms of SwiftSyntax but since `ResolvedLoc` contains a vector of `CharSourceRange`, writing a C bridging layer is a pain. So, we decided to punt this until we enable C++ interop inside the compiler.

The next steps to do then are:
- Rewrite `NameMatcher` in SwiftSyntax
- Ensure that the following issues are fixed
  - rdar://108267794
  - rdar://101556709
  - rdar://84061868
- Delete the local rename refactoring action. Renaming should always go through the sourcekitd requests
- When we remove the rename refactoring, I think we also don’t need `OldName` and `NewName` in `RenameLoc` anymore, which means that we can also remove the `StringScratchSpace` in `RenameLocs` and replace `RenameLocs` by `std::vector<RenameLoc>`.

**Comment by ahoppen** [comment]
Superseded by https://github.com/apple/swift/pull/70008

**Pull request closed without merging**